feat(api): add update vaildate email endpoint

Sboonny opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Ref #50273

I underestimated just how much work was required for this endpoint. If it's okay, I'd like to get this 'half' in, and handle the actual sending of email in another PR.

## What this does

Validation, mostly. It checks various ways in which it's possible to fail to create a request to update your email. The tricky bit was figuring out what tests to write and, while there are a lot, there are probably still gaps.

<!-- Feel free to add any additional description of changes below this line -->

Sboonny committed:
feat(api): add update vaild email endpoint

format prettier

check if the requested email isn't the same email

change the logic behind if statement

Format with prettier"

it only response with 200 code 🤔

ojeytonwilliams reviewed (changes requested):
Hey @Sboonny this endpoint does some additional checks before it updates the email address. They'll need to be recreated.

You can find them here: https://github.com/freeCodeCamp/freeCodeCamp/blob/a52b13c17ee236a4c8092d7ca013aae821877876/api-server/src/common/models/user.js#L540

Sboonny commented:
Do we need validation for the time, aka do I install `moment`, or do we add it later because in the old repo it uses `moment`?

Or do I try to add it without moment 💭

naomi-lgbt commented on api/src/email/route.ts:
```diff
@@ -0,0 +1,90 @@
+            message: Type.Literal('flash.email-valid'),
+            type: Type.Literal('success')
+          }),
+          204: Type.Object({
+            message: Type.String(),
+            type: Type.Literal('info')
+          }),
```
I don't see where we're actually sending a 204 status, but my understanding is that HTTP responses with a 204 status should not have a body.

Sboonny replied:
Oliver is taking care of the PR, because of how complex rx is, and I can't avoid it

ojeytonwilliams committed:
test: check various outcomes of update-my-email

ojeytonwilliams commented:
Still a WIP, but I added a bunch of (probably mostly correct) tests.

TODO:

- [ ] Make them green
- [ ] Figure out what to do with `requestAuthEmail`. Catch outgoing email? Mock the mail client?

ojeytonwilliams committed:
test: check that emailAuthLinkTTL is added

chore: delete duplicate route

ojeytonwilliams committed:
feat: validate is email and not existing

ojeytonwilliams committed:
test: check user record is updated correctly

ojeytonwilliams committed:
feat: reject repeated requests to make same update

ojeytonwilliams committed:
fix: newly created users should be verified

The assumption is that auth0 has provided a verified email.

ojeytonwilliams committed:
feat: allow updates to current email if unverified

ojeytonwilliams committed:
feat: reject if another user has the email address

ojeytonwilliams committed:
refactor: test for status code last

Otherwise the failing tests give you the least useful information.

fix(test): purge emailAuthLinkTTL before each test

ojeytonwilliams committed:
feat: reject too frequent requests

fix(test): too many spaces

ojeytonwilliams committed:
fix: use superRequest + fastifyTestInstance

naomi-lgbt commented on api/src/utils/default-user.ts:
```diff
@@ -0,0 +1,55 @@
+// TODO: audit this object to find out which properties need to be updated.
+export const defaultUser = {
```
```suggestion
import { user } from "@prisma/client";

export const defaultUser: Omit<user, "id" | "email"> = {
```

I actually ran into something similar in one of my own projects - keeping a default record in sync with the schema. Found that enforcing a type helped!

ojeytonwilliams replied:
Oooh, nice idea!

ojeytonwilliams committed:
fix: update status codes + serialization

refactor: drop unnecessary assertions

align to the new style of schema

rename the variable

Co-authored-by: Mrugesh Mohapatra <noreply@mrugesh.dev>

add max length to email format

Nirajn2311 commented on api/src/routes/settings.test.ts:
```diff
@@ -157,6 +176,169 @@ describe('settingRoutes', () => {
+      });
+
+      test("PUT updates the user's record in preparation for receiving auth email", async () => {
+        expect.assertions(6);
```
Since none of the `expect` statements are inside an asynchronous function, do we need this? Same goes for line 283.

naomi-lgbt replied:
It probably doesn't hurt. The test itself is async, and sometimes that does weird stuff.

ojeytonwilliams replied:
Yeah, it doesn't really hurt, but Niraj is right - we don't actually need them.

In this case I can't think of anything weird that could happen, so I'll get rid of them.

Sboonny commented on api/src/utils/default-user.ts:
```diff
@@ -4,6 +4,7 @@ import { type Prisma } from '@prisma/client';
 export const defaultUser: Omit<Prisma.userCreateInput, 'email'> = {
   about: '',
   acceptedPrivacyTerms: false,
+  badges: {},
```
```suggestion
```

Sboonny commented on api/src/utils/default-user.ts:
```diff
@@ -48,6 +48,8 @@ export const defaultUser: Omit<Prisma.userCreateInput, 'email'> = {
     showTimeLine: false
   },
   progressTimestamps: [],
+  // TODO: check what this is used for in api-server and if we need it
+  rand: 0,
```
```suggestion
```

raisedadead commented on api/src/routes/settings.ts:
```diff
@@ -77,6 +104,106 @@ export const settingRoutes: FastifyPluginCallbackTypebox = (
+
+      const isResendUpdateToSameEmail =
+        newEmail === user.newEmail?.toLowerCase();
+      const isLinkSentWithinLimit = getWaitMessage(user.emailVerifyTTL);
```
```suggestion
      const isLinkSentWithinLimitTTL = getWaitMessage(user.emailVerifyTTL);
```

ShaunSHamilton commented:
Hey @Sboonny,
I am looking to review this PR. What is the current state of it?
> If it's okay, I'd like to get this 'half' in, and handle the actual sending of email in another PR.

Is that still true?

Sboonny commented:
Actually, i have no idea, it has been open for so long that I have no context of what is going on anymore

Sboonny commented:
Shaun, sending the email isn't implemented yet, I will implement it after Niraj PR for report-user endpoint is merged, because what I will do should be the same as what he is doing with the email there (tests and allowing us to change the CC).

The validation and updating the email is finished thanks to Oliver.

ShaunSHamilton commented:
> Shaun, sending the email isn't implemented yet, I will implement it after Niraj PR for report-user endpoint is merged,

Would you like to do that in a different PR? Or, shall I review this now, and we be sure to add a `TODO(MVP):` comment to add anything extra for the email stuff?

Sboonny committed:
add todo for the email

Sboonny commented:
I would like to do it in another PR, because of the merge conflicts.

ojeytonwilliams commented:
> I would like to do it in another PR, because of the merge conflicts.

Fine by me. Are you wanting to make any further changes, @Sboonny or shall we get on with the QA?

Sboonny commented:
go for it 👍

moT01 reviewed (commented):
I think this looks good to me (aside from more conflicts) @Sboonny :+1: It's certainly a tricky endpoint, and probably one of most important. So it should maybe get an extra set of eyes or two - but I think it's working as intended. When I hit the endpoint, it sets `newEmail` on my user object - then it sounds like another PR will add the email verification, where it will set the `newEmail` to the `email` after following the link.

Sboonny committed:
fix a typo

Co-authored-by: mot01 <tmondloch01@gmail.com>

Sboonny commented:
> So it should maybe get an extra set of eyes or two - but I think it's working as intended. When I hit the endpoint, it sets `newEmail` on my user object - then it sounds like another PR will add the email verification, where it will set the `newEmail` to the `email` after following the lin

Yeab, I couldn't have worded it better, you got the jest of it 👍

Commit:
Merge branch 'main' into feat/add-update-vaild-email-endpoint

Sboonny committed:
remove duplicated import

ojeytonwilliams commented:
I'm glad to finally get this in. Thanks for your patience, @Sboonny

ShaunSHamilton commented on api/src/routes/settings.ts:
```diff
@@ -92,6 +119,107 @@ export const settingRoutes: FastifyPluginCallbackTypebox = (
+        } as const;
+      }
+
+      // ToDo(MVP): email the new email and wait user to confirm it, before we update the user schema.
+      try {
+        await fastify.prisma.user.update({
+          where: { id: req.session.user.id },
+          data: {
+            newEmail,
+            emailVerified: false,
+            emailVerifyTTL: new Date()
+          }
+        });
+
+        // TODO: combine emailVerifyTTL and emailAuthLinkTTL? I'm not sure why
+        // we need emailVeriftyTTL given that the main thing we want is to
+        // restrict the rate of attempts and the emailAuthLinkTTL already does
+        // that.
+        const tooManyRequestsMessage = getWaitMessage(user.emailAuthLinkTTL);
+
+        if (tooManyRequestsMessage) {
+          void reply.code(429);
+          return {
+            type: 'info',
+            message: tooManyRequestsMessage
+          } as const;
+        }
```
From what I can tell in the old api, we run the `tooManyRequestsMessage` before updating the email. We should do the same here.

That is, move the `if (tooManyRequestsMessage)` to be before the `fastify.prisma.user.update`
